README.md
ethstatedb
=====

Command to check Ethereum state database

[![Version](https://img.shields.io/npm/v/ethstatedb.svg)](https://npmjs.org/package/ethstatedb)
[![Downloads/week](https://img.shields.io/npm/dw/ethstatedb.svg)](https://npmjs.org/package/ethstatedb)
[![License](https://img.shields.io/npm/l/ethstatedb.svg)](https://github.com/shenghu/ethstatedb/blob/master/package.json)
[![CircleCI](https://circleci.com/gh/shenghu/ethstatedb.svg?style=svg)](https://circleci.com/gh/shenghu/ethstatedb)
[![Coverage Status](https://coveralls.io/repos/github/shenghu/ethstatedb/badge.svg?branch=master)](https://coveralls.io/github/shenghu/ethstatedb?branch=master)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)


This cli is created when I was studying Ethereum. I was curious about when and how an accout is created and maintained by Ethereum state database. From [Diving into Ethereum's world state](https://medium.com/cybermiles/diving-into-ethereums-world-state-c893102030ed), it explains that accounts are not created in state database until there is a transaction for it. So I create this small tool to verify it.  

<!-- toc -->
* [Usage](#usage)
* [Commands](#commands)
<!-- tocstop -->
# Usage
<!-- usage -->
```sh-session
$ npm install -g ethstatedb

$ ethstatedb account -d .db/chaindata/ -r 0xecc60e00b3fe5ce9f6e1a10e5469764daf51f1fe93c22ec3f9a7583a80357217 0x095e7baea6a6c7c4c2dfeb977efac326af552

Address   0x095e7baea6a6c7c4c2dfeb977efac326af552 
Balance   0                                       
Nonce     0                                       
StateRoot                                         
CodeHash 
```
<!-- usagestop -->
# Commands
<!-- commands -->
* [`ethstatedb account [ADDRESS]`](#ethstatedb-account)
* [`ethstatedb help`](#ethstatedb-help)

## `ethstatedb account [ADDRESS]`

Retrieve account information from Ethereum state database

```
USAGE
  $ ethstatedb account ADDRESS

ARGUMENTS
  ADDRESS  public key of an account, e.g. 
           0x7C476EA4e20D6d49820Ee810C0C474ED11b3f527

OPTIONS
  -d, --dbdir=dbdir    (required) path of Ethereum state database, e.g. 
                       ethereum/geth/chaindata

  -h, --help           show CLI help

  -r, --root=root      (required) value of stateRoot

  --skip-copy          skip copying database to .ethstatedb. By default, 
                       database is copied to avoid messing the original 
                       database.

  --work-dir=work-dir  directory to pull down database

EXAMPLE
  $ ethstatedb account --dbdir /home/data/ethereum-node1/geth/chaindata --root 
  0x595e3a0db7f38600770a267b12b67a8da1c202fef4b6c21e60fdb92477d89753 
  0x7C476EA4e20D6d49820Ee810C0C474ED11b3f527
```

## `ethstatedb help`

Display help for ethstatedb

```
Command to check Ethereum state database

VERSION
  ethstatedb/1.0.5 darwin-x64 node-v10.15.3

USAGE
  $ ethstatedb [COMMAND]

COMMANDS
  account  load information of an account
  help     display help for ethstatedb
```
<!-- commandsstop -->

# More information
1. When playing with data generated by geth, make sure to stop geth so that it flushes all data to database. This command by default copies the data to a temp location, i.e. `.ethstatedb`, to avoid messing up the original database. However, you can use `--skip-copy` to use original database. 
2. To get stateRoot, use `geth attach ipc://path/to/geth.ipc` to connect to a node. Then use `eth.getBlock(eth.blockNumber).stateRoot` to get stateRoot for the latest block. You can change `eth.blockNumber` to be number of the specific block that you are interested in. 

# References
1. [ethereum_database_research_and_testing](https://github.com/tpmccallum/ethereum_database_research_and_testing/blob/master/leveldb/leveldb.md)
2. [ethereumjs-vm](https://github.com/ethereumjs/ethereumjs-vm)
3. [Diving into Ethereum's world state](https://medium.com/cybermiles/diving-into-ethereums-world-state-c893102030ed)


CHANGELOG.md
## [1.0.8](https://github.com/shenghu/ethstatedb/compare/v1.0.7...v1.0.8) (2019-07-23)


### Bug Fixes

* security update for dependencies ([7c127e5](https://github.com/shenghu/ethstatedb/commit/7c127e5))

## [1.0.7](https://github.com/shenghu/ethstatedb/compare/v1.0.6...v1.0.7) (2019-05-28)


### Bug Fixes

* remove formatBlockHeader ([1a1c281](https://github.com/shenghu/ethstatedb/commit/1a1c281))

## [1.0.6](https://github.com/shenghu/ethstatedb/compare/v1.0.5...v1.0.6) (2019-05-24)


### Bug Fixes

* Update README ([0b0db0a](https://github.com/shenghu/ethstatedb/commit/0b0db0a))

## [1.0.5](https://github.com/shenghu/ethstatedb/compare/v1.0.4...v1.0.5) (2019-05-24)


### Bug Fixes

* bin/run is not found ([eb2be91](https://github.com/shenghu/ethstatedb/commit/eb2be91))

## [1.0.4](https://github.com/shenghu/ethstatedb/compare/v1.0.3...v1.0.4) (2019-05-24)


### Bug Fixes

* add --skip-copy to avoid copy state db ([34069a5](https://github.com/shenghu/ethstatedb/commit/34069a5))

## [1.0.3](https://github.com/shenghu/ethstatedb/compare/v1.0.2...v1.0.3) (2019-05-23)


### Bug Fixes

* add changelog ([45cd0c3](https://github.com/shenghu/ethstatedb/commit/45cd0c3))


test/utils/run-blockchain.ts
import Account from 'ethereumjs-account'
import * as ethereumUtil from 'ethereumjs-util'
import * as fs from 'fs-extra'
import leveldown from 'leveldown'
import levelup from 'levelup'
import * as util from 'util'

const ethjsUtil = require('ethjs-util')
const Trie = require('merkle-patricia-tree/secure')
const Block = require('ethereumjs-block')
const Blockchain = require('ethereumjs-blockchain')
const VM = require('ethereumjs-vm')
const level = require('level')

function format(a: any, toZero?: any, isHex?: boolean) {
  if (a === '') {
    return Buffer.alloc(0)
  }

  if (a.slice && a.slice(0, 2) === '0x') {
    a = a.slice(2)
    if (a.length % 2) a = '0' + a
    a = Buffer.from(a, 'hex')
  } else if (!isHex) {
    a = Buffer.from(new ethereumUtil.BN(a).toArray())
  } else {
    if (a.length % 2) a = '0' + a
    a = Buffer.from(a, 'hex')
  }

  if (toZero && a.toString('hex') === '') {
    a = Buffer.from([0])
  }

  return a
}

async function setupPreConditions(state: any, testData: any) {
  const keysOfPre = Object.keys(testData.pre)

  return Promise.all(keysOfPre.map(async (key: any) => {
    const acctData = testData.pre[key]
    const account = new Account()

    account.nonce = format(acctData.nonce)
    account.balance = format(acctData.balance)

    const storageTrie = state.copy()
    storageTrie.root = null

    const keys = Object.keys(acctData.storage)
    await Promise.all(keys.map(async (key: any) => {
      let val = acctData.storage[key]
      val = ethereumUtil.rlp.encode(Buffer.from(val.slice(2), 'hex'))
      key = ethereumUtil.setLength(Buffer.from(key.slice(2), 'hex'), 32)

      return util.promisify(storageTrie.put)(key, val)
    }))

    const codeBuf = Buffer.from(acctData.code.slice(2), 'hex')
    await new Promise((resolve, reject) => {
      account.setCode(state, codeBuf, (err: any, codeHash: Buffer) => {
        if (err) return reject(err)
        resolve(codeHash)
      })
    })

    account.stateRoot = storageTrie.root
    if (testData.exec && key === testData.exec.address) {
      testData.root = storageTrie.root
    }
    return util.promisify(state.put)(Buffer.from(ethjsUtil.stripHexPrefix(key), 'hex'), account.serialize())
  }))
}

export const enum SupportedHardforkType {
  byzantium = 'byzantium'
}

export interface BlockChainConfig {
  dbPath: string,
  validate: boolean,
  cachePath: string,
  hardfork: SupportedHardforkType
  data: any
  reset?: boolean
}

export async function runBlockchain(config: BlockChainConfig) {
  const hardfork = config.hardfork
  const validate = config.validate
  const data = config.data

  if (config.reset) {
    await Promise.all([config.dbPath, config.cachePath].map(async path => {
      if (path) {
        await fs.emptyDir(path)
      }
    }))
  }

  await Promise.all([config.dbPath, config.cachePath].map(async dir => {
    if (dir) {
      await fs.ensureDir(dir)
    }
  }))

  const db = levelup(leveldown(config.dbPath))
  const blockchain = new Blockchain({db, hardfork, validate})

  if (validate)
    blockchain.ethash.cacheDB = level(config.cachePath)

  const state = new Trie(db)
  const vm = new VM({state, blockchain, hardfork})
  data.homestead = true

  vm.on('beforeTx', (tx: any) => {
    tx._homestead = true
  })

  vm.on('beforeBlock', (block: any) => {
    block.header.isHomestead = function () {
      return true
    }
  })

  const genesisBlock = new Block({hardfork})

  // Setup pre state
  await setupPreConditions(state, data)

  // Create genesis block
  await new Promise((resolve, reject) => {
    genesisBlock.header = new Block.Header(data.genesisBlockHeader,
      {hardfork})
    blockchain.putGenesis(genesisBlock, (err: any, value: any) => {
      if (err) reject(err)

      resolve(value)
    })
  })

  // add some following blocks
  for (const raw of data.blocks) {
    const block = new Block(Buffer.from(raw.rlp.slice(2), 'hex'))

    // forces the block into thinking they are homestead
    block.header.isHomestead = () => {
      return true
    }

    block.uncleHeaders.forEach((uncle: any) => {
      uncle.isHomestead = function () { return true }
    })

    await new Promise((resolve, reject) => {
      blockchain.putBlock(block, (err: any, result: any) => {
        if (err) reject(err)
        resolve(result)
      })
    })
  }

  // Run blockchain
  await new Promise((resolve, reject) => {
    vm.runBlockchain((err: any, value: any) => {
      if (err) reject(err)

      resolve(value)
    })
  })

  return vm
}


test/commands/account.test.ts
import {expect, test} from '@oclif/test'
import * as fs from 'fs-extra'
import * as os from 'os'

import {runBlockchain, SupportedHardforkType} from '../utils/run-blockchain'

const testData = require('../fixtures/test-blockchain-data.json')

describe('account', () => {
  before(async () => {
    const evm = await runBlockchain({
      dbPath: '.db/chaindata',
      cachePath: '.db/cachedata',
      validate: true,
      data: testData,
      hardfork: SupportedHardforkType.byzantium,
      reset: true
    })
  
    await new Promise((resolve, reject) => {
      evm.blockchain.db.close((err: Error, result: any) => {
        if (err) reject(err)
  
        resolve(result)
      })
    })
  })

  test
    .stdout()
    .do(() => {
      fs.copySync('.db/chaindata', '.db/chaindata-1')
    })
    .command(['account',
      '--dbdir', '.db/chaindata-1',
      '--root', '0xecc60e00b3fe5ce9f6e1a10e5469764daf51f1fe93c22ec3f9a7583a80357217',
      '--work-dir', '.ethstatedb-1',
      '0x095e7baea6a6c7c4c2dfeb977efac326af552d87'])
    .it('runs account', async ctx => {
      const result: any = {}
      ctx.stdout.split(os.EOL).map((element: string) => {
        const tokens = element.trim().split(/\s+/)
        if (tokens.length > 1) {
          result[tokens[0]] = tokens[1]
        }
      })

      expect(result.Balance).to.equal('10')
    })

  test
    .stdout()
    .do(() => {
      fs.copySync('.db/chaindata', '.db/chaindata-2')
      fs.removeSync('.ethstatedb')
    })
    .command(['account',
      '--dbdir', '.db/chaindata-2',
      '--root', '0xecc60e00b3fe5ce9f6e1a10e5469764daf51f1fe93c22ec3f9a7583a80357217',
      '--skip-copy',
      '0x095e7baea6a6c7c4c2dfeb977efac326af552d87'])
    .it('runs account --skip-copy', async ctx => {
      const result: any = {}
      ctx.stdout.split(os.EOL).map((element: string) => {
        const tokens = element.trim().split(/\s+/)
        if (tokens.length > 1) {
          result[tokens[0]] = tokens[1]
        }
      })

      expect(result.Balance).to.equal('10')
      expect(await fs.pathExists('.ethstatedb')).to.be.false
    })
})


src/commands/account.ts
import {Command, flags} from '@oclif/command'
import cli from 'cli-ux'
import EAccount from 'ethereumjs-account'
import {BN, KECCAK256_NULL_S, KECCAK256_RLP_S} from 'ethereumjs-util'
import * as fs from 'fs-extra'

export default class Account extends Command {
  static description = 'load information of an account'

  static examples = [
    '$ ethstatedb account --dbdir /home/data/ethereum-node1/geth/chaindata --root 0x595e3a0db7f38600770a267b12b67a8da1c202fef4b6c21e60fdb92477d89753 0x7C476EA4e20D6d49820Ee810C0C474ED11b3f527',
  ]

  static flags = {
    help: flags.help({char: 'h'}),
    dbdir: flags.string({
      char: 'd',
      required: true,
      description:
        'path of Ethereum state database, e.g. ethereum/geth/chaindata',
    }),
    root: flags.string({
      char: 'r',
      required: true,
      description: 'value of stateRoot',
    }),
    'skip-copy': flags.boolean({
      required: false,
      description:
        'skip copying database to .ethstatedb. By default, database is copied to avoid messing the original database.',
    }),
    'work-dir': flags.string({
      required: false,
      description: 'directory to pull down database',
    }),
  }

  static args = [
    {
      name: 'address',
      required: true,
      description:
        'public key of an account, e.g. 0x7C476EA4e20D6d49820Ee810C0C474ED11b3f527',
    },
  ]

  async run() {
    const Trie = require('merkle-patricia-tree/secure')
    const level = require('level')

    const {args, flags} = this.parse(Account)

    let workDir = flags.dbdir
    if (!flags['skip-copy']) {
      workDir = flags['work-dir'] || '.ethstatedb'
      await fs.emptyDir(workDir)
      await fs.copy(flags.dbdir, workDir)
    }

    const db = level(workDir)
    const trie = new Trie(db, flags.root)

    const found = await new Promise((resolve, reject) => {
      trie.checkRoot(flags.root, (err: any, value: boolean) => {
        if (err) return reject(err)

        resolve(value)
      })
    })

    if (!found) return this.warn(`stateRoot (${flags.root}) is not found in db`)

    const account: EAccount = await new Promise((resolve, reject) => {
      trie.get(args.address, (err: Error, raw: any) => {
        if (err) return reject(err)

        return resolve(new EAccount(raw))
      })
    })

    this.log('')
    this.print(args.address, account)
  }

  print(address: string, account: EAccount) {
    cli.table(
      [
        {name: 'Address', value: address},
        {
          name: 'Balance',
          value: `${new BN(account.balance).toString()}`,
        },
        {name: 'Nonce', value: `${new BN(account.nonce).toString()}`},
        {
          name: 'StateRoot',
          value: `${
            account.stateRoot.toString('hex') === KECCAK256_RLP_S
              ? ''
              : '0x' + account.stateRoot.toString('hex')
          }`,
        },
        {
          name: 'CodeHash',
          value: `${
            account.codeHash.toString('hex') === KECCAK256_NULL_S
              ? ''
              : '0x' + account.codeHash.toString('hex')
          }`,
        },
      ],
      {name: {}, value: {}},
      {printLine: this.log, 'no-header': true},
    )
  }
}
